common/src/WooCommerce.php

```diff
@@ -464,6 +465,100 @@ public function getManualSyncNonce(): string {
         return $this->plugin->getOptionName('manual-sync-data');
     }
 
+    public function getProductKeysAction(): string {
+        return $this->plugin->getOptionName('product_keys');
+    }
+
+    public function getProductKeys() {
+        $selected_key = $_GET['selected_key'];
+        wp_send_json([
+            'status' => true,
+            'data' => array_map(
+                function ($value) use ($selected_key) {
+                    $option_value = $value['type'] . $value['name'];
+                    return [
+                        'option_value' => $option_value,
+                        'label' => sprintf('%s (e.g. "%s")', $value['name'], $value['example_value']),
+                        'suggested' => $this->isValidGtin($value['example_value']),
+                        'selected' => $option_value == $selected_key,
+                    ];
+                },
+                array_merge($this->getProductMetaKeys(), $this->getCustomAttributes())
+            ),
+        ]);
+        wp_die();
+    }
+
+    public function getProductMetaKeys(): array {
+        global $wpdb;
+        $meta_keys=[];
+        $meta_keys_data = $wpdb->get_col("
+            SELECT DISTINCT(pm.meta_key)
+            FROM {$wpdb->posts} p
+            LEFT JOIN {$wpdb->postmeta} pm ON p.ID = pm.post_id
+            WHERE
+                p.post_type = 'product'
+                AND pm.meta_key <> ''
+                AND pm.meta_value <> ''
+        ");
+        foreach ($meta_keys_data as $value) {
+            $meta_keys[$value] =[
+                'type' => 'meta_key',
+                'name' => $value,
+                'example_value' => substr($this->getMetaValue($value), 0, 15),
+            ];
+        }
+        return $meta_keys;
+    }
+
+    private function getMetaValue(string $meta_key) {
+        global $wpdb;
+        $sql = "
+            SELECT meta.meta_value
+            FROM {$wpdb->postmeta} meta
+            WHERE meta.meta_key = %s
+            AND meta.meta_value <> ''
+            ORDER BY meta.meta_id DESC
+            LIMIT 1;
+        ";
+        return $wpdb->get_var($wpdb->prepare($sql, $meta_key));
+    }
+
+    private function getCustomAttributes(): array {
+        global $wpdb;
+        $custom_attributes = [];
+        $sql = "
+            SELECT meta.meta_id, meta.meta_key as name, meta.meta_value 
+            FROM {$wpdb->postmeta} meta
+            JOIN {$wpdb->posts} posts
+            ON meta.post_id = posts.id 
+            WHERE posts.post_type = 'product' 
+            AND meta.meta_key='_product_attributes'
+            ORDER BY posts.id DESC
+            LIMIT 1000;
+        ";
+
+        $data = $wpdb->get_results($sql);
+        foreach ($data as $value) {
+            $product_attr = unserialize($value->meta_value);
```

Comment:
To make this a little safer (unserialization can cause errors with certain values, non-existent classes, etc.), you can do this:

~~~php
try {
  $product_attr = unserialize($value->meta_value, ['allowed_classes' => false]);
} catch (\Throwable $e) {
  continue;
}
~~~

* Don't prevent unserialization of classes.
* If an error occurs during unserialization, just ignore it.